wvdecrypter/wvdecrypter.cpp

```diff
@@ -1166,99 +1253,135 @@ AP4_Result WV_CencSingleSampleDecrypter::DecryptSampleData(AP4_UI32 pool_id,
-      numCipherBytes += subsamples[i].cipher_bytes;
+      numCipherBytes += bytes_of_encrypted_data[i];
     }
-    if (numCipherBytes > 0)
-    {
-      cdm_in.data = data_in.GetData();
-      cdm_in.data_size = data_in.GetDataSize();
-      cdm_in.num_subsamples = subsample_count;
-    }
-    else
+
+    if (numCipherBytes == 0)
     {
-      memcpy(data_out.UseData(), data_in.GetData(), data_in.GetDataSize());
+      data_out.AppendData(data_in.GetData(), data_in.GetDataSize());
       return AP4_SUCCESS;
     }
-  }
-
-  cdm_in.iv = iv;
-  cdm_in.iv_size = 16; //Always 16, see AP4_CencSingleSampleDecrypter declaration.
-  cdm_in.key_id = fragInfo.key_;
-  cdm_in.key_id_size = 16;
-  cdm_in.subsamples = subsamples.data();
-  cdm_in.encryption_scheme = media::ToCdmEncryptionScheme(m_EncryptionMode);
-  cdm_in.timestamp = 0;
-  cdm_in.pattern = { m_CryptBlocks, m_SkipBlocks };
-
-  CdmBuffer buf((useSingleDecrypt) ? &decrypt_out_ : &data_out);
-  CdmDecryptedBlock cdm_out;
-  cdm_out.SetDecryptedBuffer(&buf);
 
-  //LICENSERENEWAL: 
-  CheckLicenseRenewal();
-  cdm::Status ret = drm_.GetCdmAdapter()->Decrypt(cdm_in, &cdm_out);
+    cdm::InputBuffer_2 cdm_in;
+    SetInput(cdm_in, data_in, subsample_count, iv, fragInfo, subsamples);
+    data_out.SetDataSize(data_in.GetDataSize());
+    CdmBuffer buf{&data_out};
+    CdmDecryptedBlock cdm_out;
+    cdm_out.SetDecryptedBuffer(&buf);
 
-  if (ret == cdm::Status::kSuccess && useSingleDecrypt)
-  {
-    size_t absPos = 0, cipherPos = 0;
-    for (unsigned int i(0); i < subsample_count; ++i)
+    CheckLicenseRenewal();
+    ret = drm_.GetCdmAdapter()->Decrypt(cdm_in, &cdm_out);
+    if (ret != cdm::Status::kSuccess)
     {
-      memcpy(data_out.UseData() + absPos, data_in.GetData() + absPos, bytes_of_cleartext_data[i]);
-      absPos += bytes_of_cleartext_data[i];
-      memcpy(data_out.UseData() + absPos, decrypt_out_.GetData() + cipherPos, bytes_of_encrypted_data[i]);
-      absPos += bytes_of_encrypted_data[i], cipherPos += bytes_of_encrypted_data[i];
+      LogDecryptError(ret, fragInfo.key_);
     }
+    return (ret == cdm::Status::kSuccess) ? AP4_SUCCESS : AP4_ERROR_INVALID_PARAMETERS;
   }
 
-  if (ret != cdm::Status::kSuccess)
+  // We can only decrypt with subsamples set to 1
+  // This must be handled differently for CENC and CBCS
+  // CENC:
+  // CDM should get 1 block of encrypted data per sample, encrypted data
+  // from all subsamples should be formed into a contiguous block.
+  // Even if there is only 1 subsample, we should remove cleartext data
+  // from it before passing to CDM.
+  // CBCS:
+  // Due to the nature of this cipher subsamples must be decrypted separately
+  else
   {
-    char buf[36]; buf[32] = 0;
-    AP4_FormatHex(fragInfo.key_, 16, buf);
-    LOG::LogF(SSDDEBUG, "Decrypt failed with error: %d and key: %s", ret, buf);
-  }
+    const size_t iterations{useCbcDecrypt ? subsample_count : 1};
```

Comment:
has been discussed recently on slack that for better reading is some cases we can still use `=` instead of brackets (yes docs should be updated....)
```suggestion
    const size_t iterations = useCbcDecrypt ? subsample_count : 1;
```

Reply:
since subsample_count is `unsigned int` and there is no places where `size_t` is used i think its better change all size_t vars with same type or better `uint32_t`
this also on all new pack methods

Reply:
Since Bento4 uses unsigned int I'll switch to that